File: Scala/src/main/scala/TextbookTrack/Chapter01/BA1I.scala
package TextbookTrack.Chapter01

object BA1I {
  import scala.math.Integral.Implicits._

  private val Nucleotides: Vector[Char] = Vector('A', 'C', 'G', 'T')
  private val NucleotideOrder: Map[Char, Int] = Map('A' -> 0, 'C' -> 1, 'G' -> 2, 'T' -> 3)

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def readParameters(line: String): (Int, Int) = convertToIntList(line) match {
    case List(k, d) => (k, d)
    case _ => throw new Exception("Unexpected input data format.")
  }

  private def patternToNumber(pattern: String): Int =
    pattern.foldLeft(0)((acc, nucleotide) => 4 * acc + NucleotideOrder(nucleotide))

  private def numberToPattern(encoding: Int, k: Int): String = {
    val (pattern, _): (List[Char], Int) = (0 until k).foldLeft((List.empty[Char], encoding)) {
      case ((acc, code), _) =>
        val (nextCode, remainder): (Int, Int) = code /% 4
        (Nucleotides(remainder) :: acc, nextCode)
    }
    pattern.mkString
  }

  private def calcFrequenciesForApproximateOccurrences(text: String, k: Int, d: Int): Array[Int] = {
    val frequencies: Array[Int] = Array.fill(math.pow(4, k).toInt)(0)
    val kMerEncodings: Iterator[Int] = text.sliding(k).flatMap{ pattern =>
      val neighbours: Set[String] = generateNeighbourhood(pattern, d)
      neighbours.map(patternToNumber)
    }
    kMerEncodings.foreach(code => frequencies(code) += 1)
    frequencies
  }

  private def generateNeighbourhood(text: String, d: Int): Set[String] = {
    def calcImmediateNeighbours(pattern: String): Set[String] =
      pattern.view.zipWithIndex.foldLeft(Set.empty[String]){
        case (neighbours, (letter, ix)) =>
          neighbours ++ Nucleotides.filterNot(_ == letter).map(nucleotide => pattern.updated(ix, nucleotide)).toSet
      }

    (0 until d).foldLeft(Set(text)) {
      case (neighbourhood, _) => neighbourhood ++ neighbourhood.flatMap(calcImmediateNeighbours)
    }
  }

  def mostFrequentApproximateKMers(text: String, k: Int, d: Int): List[String] = {
    val frequencies: Array[Int] = calcFrequenciesForApproximateOccurrences(text, k, d)
    val maxOccurrence: Int = frequencies.max
    frequencies
      .iterator
      .zipWithIndex
      .filter{ case (freq, _) => freq == maxOccurrence }
      .map{ case (_, ix) => numberToPattern(ix, k) }
      .toList
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next()
    val (k, d): (Int, Int) = readParameters(reader.next())
    val result: List[String] = mostFrequentApproximateKMers(text, k, d)
    println(result.mkString(" "))
  }
}

File: Julia/textbook_track/chapter01/ba1i.jl
# Find the Most Frequent Words with Mismatches in a String

const NUCLEOTIDES = ('A', 'C', 'G', 'T')
const NUCLEOTIDE_ORDER = Dict{Char, Int64}('A' => 0, 'C' => 1, 'G' => 2, 'T' => 3)


convert_to_intlist(line) = map(x -> parse(Int, x), split(line))


pattern_to_number(pattern) = foldl((acc, nucleotide) -> 4 * acc + NUCLEOTIDE_ORDER[nucleotide], pattern; init=0)


function number_to_pattern(encoding, k)
    pattern = Char[]
    for _ in 1:k
        encoding, remainder = divrem(encoding, 4)
        push!(pattern, NUCLEOTIDES[remainder+1])
    end
    join(reverse(pattern))
end


function calc_frequencies_for_approximate_occurrences(text, k, d)
    frequencies = zeros(Int32, 4 ^ k)
    for ix in 1:(length(text)-k)
        exact_pattern = text[ix:ix+k-1]
        for pattern in generate_neighbourhood(exact_pattern, d)
            code = pattern_to_number(pattern)
            frequencies[code+1] += 1
        end
    end
    frequencies
end


function calc_immediate_neighbours(pattern)
    neighbours = Set{String}()
    for (ix, letter) in enumerate(pattern)
        for nucleotide in NUCLEOTIDES
            if nucleotide != letter
                pattern_array = collect(pattern)
                pattern_array[ix] = nucleotide
                push!(neighbours, join(pattern_array))
            end
        end
    end
    neighbours
end


function generate_neighbourhood(text, d)
    neighbourhood = Set{String}([text])
    for _ in 1:d
        neighbours_of_neighbours = Set{String}()
        for pattern in neighbourhood
            immediate_neighbours = calc_immediate_neighbours(pattern)
            union!(neighbours_of_neighbours, immediate_neighbours)
        end
        union!(neighbourhood, neighbours_of_neighbours)
    end
    neighbourhood
end


function most_frequent_approximate_k_mers(text, k, d)
    frequencies = calc_frequencies_for_approximate_occurrences(text, k, d)
    max_occurrence = maximum(frequencies)
    patterns = String[]
    for (ix, freq) in enumerate(frequencies)
        if freq == max_occurrence
            pattern = number_to_pattern(ix - 1, k)
            push!(patterns, pattern)
        end
    end
    patterns
end


function main()
    text = readline()
    k, d = convert_to_intlist(readline())
    result = most_frequent_approximate_k_mers(text, k, d)
    println(join(result, " "))
end


main()
